### ISSUE TITLE
Upload new Hackage version compatible with GHC 9.2.5

### ISSUE THREAD

**Issue by ltbinsbe** [conal/TypeCompose#12]
```
Building library for TypeCompose-0.9.14..
[ 1 of 10] Compiling Control.Instances ( src/Control/Instances.hs, dist/build/Control/Instances.o, dist/build/Control/Instances.dyn_o )
[ 2 of 10] Compiling Data.Bijection   ( src/Data/Bijection.hs, dist/build/Data/Bijection.o, dist/build/Data/Bijection.dyn_o )
[ 3 of 10] Compiling Control.Compose  ( src/Control/Compose.hs, dist/build/Control/Compose.o, dist/build/Control/Compose.dyn_o )

src/Control/Compose.hs:221:22: warning: [-Wstar-is-type]
    Using ‘*’ (or its Unicode variant) to mean ‘Data.Kind.Type’
    relies on the StarIsType extension, which will become
    deprecated in the future.
    Suggested fix: use ‘Type’ from ‘Data.Kind’ instead.
    |
221 | newtype ((g :: k2 -> *) :. (f :: k1 -> k2)) (a :: k1)
    |                      ^

src/Control/Compose.hs:748:10: warning: [-Wsimplifiable-class-constraints]
    • The constraint ‘Show (f a, g a)’ matches
        instance (Show a, Show b) => Show (a, b) -- Defined in ‘GHC.Show’
      This makes type inference for inner bindings fragile;
        either use MonoLocalBinds, or simplify it using the instance
    • In the instance declaration for ‘Show ((f :*: g) a)’
    |
748 | instance (Show (f a, g a)) => Show ((f :*: g) a) where
    |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

src/Control/Compose.hs:751:10: warning: [-Wsimplifiable-class-constraints]
    • The constraint ‘Eq (f a, g a)’ matches
        instance (Eq a, Eq b) => Eq (a, b)
          -- Defined in ‘ghc-prim:GHC.Classes’
      This makes type inference for inner bindings fragile;
        either use MonoLocalBinds, or simplify it using the instance
    • In the instance declaration for ‘Eq ((f :*: g) a)’
    |
751 | instance (Eq (f a, g a)) => Eq ((f :*: g) a) where
    |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

src/Control/Compose.hs:754:10: warning: [-Wsimplifiable-class-constraints]
    • The constraint ‘Ord (f a, g a)’ matches
        instance (Ord a, Ord b) => Ord (a, b)
          -- Defined in ‘ghc-prim:GHC.Classes’
      This makes type inference for inner bindings fragile;
        either use MonoLocalBinds, or simplify it using the instance
    • In the instance declaration for ‘Ord ((f :*: g) a)’
    |
754 | instance (Ord (f a, g a)) => Ord ((f :*: g) a) where
    |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[ 4 of 10] Compiling Data.Lambda      ( src/Data/Lambda.hs, dist/build/Data/Lambda.o, dist/build/Data/Lambda.dyn_o )
[ 5 of 10] Compiling Data.Pair        ( src/Data/Pair.hs, dist/build/Data/Pair.o, dist/build/Data/Pair.dyn_o )

src/Data/Pair.hs:98:10: error:
    • Non type-variable argument in the constraint: Monoid_f (Flip j o)
      (Use FlexibleContexts to permit this)
    • In the instance declaration for ‘Pair (Flip j o)’
   |
98 | instance (Arrow j, Monoid_f (Flip j o)) =>
   |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
Error: cabal: Failed to build TypeCompose-0.9.14
```

**Comment by conal** [comment]
Thanks for the error report. This library is very old, and mostly superseded (e.g., by `GHC.Generics`), afaik. Are you using it?

**Comment by ltbinsbe** [comment]
Yes indeed, as part of the `gll` library which itself is used in several of my other projects. Would you suggest migrating away from TypeCompose? I have no idea how involved that would be.

Many thanks for your quick response.
________________________________

**Comment by conal** [comment]
> Would you suggest migrating away from TypeCompose?

It's worth exploring. I wrote `TypeCompose` to capture some useful patterns involving functor algebra that came up often for me. Later, `GHC.Generics` provided much of that functionality and gave it just enough compiler support ("`deriving Generic`") to be very convenient. I switched to that functionality for my own work, e.g., [*Generic functional parallel algorithms: Scan and FFT*](http://conal.net/papers/generic-parallel-functional). (Now I'm doing something analogous in Agda, which is a much more natural setting for generic/polytypic programming.)

> I have no idea how involved that would be.

I just looked at the `gll` source. It appears to use only `OO` from `Control.Compose`:

``` agda
-- | Composition of type constructors: unary with binary.  Called
-- "StaticArrow" in [1].
newtype OO f j a b = OO { unOO :: f (a `j` b) }

#if __GLASGOW_HASKELL__ >= 609
instance (Applicative f, Category cat) => Category (OO f cat) where
  id          = OO (pure id)
  OO g . OO h = OO (liftA2 (.) g h)
#endif

instance (Applicative f, Arrow arr) => Arrow (OO f arr) where
#if __GLASGOW_HASKELL__ < 609
  OO g >>> OO h = OO (liftA2 (>>>) g h)
#endif
  arr           = OO . pure . arr
  first (OO g)  = OO (liftA first g)
```

That `OO` type construction (which I'd forgotten) is *not* in `GHC.Generics`. I think that "\[1\]" is meant to be [Applicative Programming with Effects](https://www.staff.city.ac.uk/~ross/papers/Applicative.html). Sure enough, this construction is [in `Control.Arrow.Transformer.Static`](https://hackage.haskell.org/package/arrows-0.4.4.2/docs/Control-Arrow-Transformer-Static.html#t:StaticArrow).

You're most welcome to copy the `OO` code above and use it in `gll`. Better for you, however, might be to use `StaticArrow` instead so you have a little less code to maintain.

Good luck, and please let me know how it comes out.

-- Conal

**Comment by ltbinsbe** [comment]
Many thanks for taking the effort to investigate the matter so thoroughly!

A first attempt revealed that by using StaticArrow I will now need to give Arrow and Category instances for at least one of my types that I did not need for OO.
I need to dive deeper to see if this raises any problems.
________________________________

**Comment by ltbinsbe** [comment]
For the time being I have followed your first suggestion and copied over the needed definition of `OO`

**Comment by conal** [comment]
> A first attempt revealed that by using StaticArrow I will now need to give Arrow and Category instances for at least one of my types that I did not need for OO.

Odd. On inspection, `StaticArrow` and `OO` appear to have exactly the same requirements for constructing and for `Category` and `Arrow` instances. ` ¯\_(ツ)_/¯`

**Comment by conal** [comment]
If you figure out the source of the difference, I'd love to hear.

**Comment by ltbinsbe** [comment]
The missing instance error was triggered by the use of `unWrap` instead of `unOO`

**Comment by conal** [comment]
Oho! Thanks for the explanation. A questionable choice for the type of `unWrap`.

It sounds like you were not using the `Category` or `Arrow` instances for `OO` (which would have made the same demand). In that case, all you'd need to copy is the `OO` definition itself, which is probably a good way to go.

**Comment by ltbinsbe** [comment]
Yes, I went this route. Just published a new version of the `gll` library. Now updating the packages downstream :crossed_fingers: 

Thanks again for your help and interest.

**Comment by Bodigrim** [comment]
As a Hackage trustee I made revisions to prohibit unbuildable configurations:
https://hackage.haskell.org/package/TypeCompose-0.9.13/revisions/
https://hackage.haskell.org/package/TypeCompose-0.9.14/revisions/

**Comment by smunix** [comment]
> As a Hackage trustee I made revisions to prohibit unbuildable configurations: https://hackage.haskell.org/package/TypeCompose-0.9.13/revisions/ https://hackage.haskell.org/package/TypeCompose-0.9.14/revisions/

The build issues reported here with ghc925 were already fixed, see #11. 

@conal I think it's still worth pushing a new version (0.9.15?) to hackage, as multiple other packages have a transitive dependency landing on `TypeCompose`. What do you think?

**Comment by conal** [comment]
> > As a Hackage trustee I made revisions to prohibit unbuildable configurations: https://hackage.haskell.org/package/TypeCompose-0.9.13/revisions/ https://hackage.haskell.org/package/TypeCompose-0.9.14/revisions/
> 
> The build issues reported here with ghc925 were already fixed, see #11.
> 
> @conal I think it's still worth pushing a new version (0.9.15?) to hackage, as multiple other packages have a transitive dependency landing on `TypeCompose`. What do you think?

@smunix I think it would be healthier for the community to drop TypeCompose for more modern, complete, and robust alternatives. By *not* updating TypeCompose, we're more likely to evaluate each current use as the current situation becomes problematic for those uses, as with @ltbinsbe above. If there are any remaining wise uses of TypeCompose, the strategy I'm suggesting will surface them.

**Comment by conal** [comment]
Do we (Hackage users) have successful and recommended strategies for deprecating currently used packages?

**Comment by smunix** [comment]
> @smunix I think it would be healthier for the community to drop TypeCompose for more modern, complete, and robust alternatives. By _not_ updating TypeCompose, we're more likely to evaluate each current use as the current situation becomes problematic for those uses, as with @ltbinsbe above. If there are any remaining wise uses of TypeCompose, the strategy I'm suggesting will surface them.

Noted, thanks. It turns out the `O` composition is equivalent to `Data.Functor.Compose` in `base`; so I went ahead and removed `TypeCompose` from `spatial-math`. See the PR above ^

**Comment by conal** [comment]
> Noted, thanks. It turns out the `O` composition is equivalent to `Data.Functor.Compose` in `base`; so I went ahead and removed `TypeCompose` from `spatial-math`. See the PR above ^

Perfect! TypeCompose pre-dates `Data.Functor.Compose` and `GHC.Generics`, and I switched over to the modern replacements for my own work many years ago.

Thanks for letting me know of the happy outcome, @smunix!

**Comment by hsenag** [comment]
`Flip` appears to have a [suitable replacement](https://hackage.haskell.org/package/bifunctors-5.6.1/docs/Data-Bifunctor-Flip.html) in the `bifunctors` package.
